lib/wasi/src/state/pipe.rs

```diff
@@ -93,26 +234,108 @@ impl WasiPipe {
     }
 }
 
+impl Write for WasiPipe {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let buf_len = buf.len();
+        let tx = self.tx.lock().unwrap();
+        tx.send(buf.to_vec())
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, format!("{e}")))?;
+        Ok(buf_len)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Seek for WasiPipe {
+    fn seek(&mut self, _: SeekFrom) -> io::Result<u64> {
+        Ok(0)
+    }
+}
+
 impl Read for WasiPipe {
     fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
         loop {
             if let Some(inner_buf) = self.read_buffer.as_mut() {
                 let buf_len = inner_buf.len();
                 if buf_len > 0 {
-                    let mut reader = inner_buf.as_ref();
-                    let read = reader.read(buf).map(|_| buf_len as usize)?;
-                    inner_buf.advance(read);
-                    return Ok(read);
+                    if inner_buf.len() > buf.len() {
+                        let mut reader = inner_buf.as_ref();
+                        let read = reader.read_exact(buf).map(|_| buf.len())?;
+                        inner_buf.advance(read);
+                        return Ok(read);
+                    } else {
+                        let mut reader = inner_buf.as_ref();
+                        let read = reader.read(buf).map(|_| buf_len as usize)?;
+                        inner_buf.advance(read);
+                        return Ok(read);
+                    }
+                } else {
+                    return Ok(0);
                 }
             }
             let rx = self.rx.lock().unwrap();
-            let data = rx.recv().map_err(|_| {
-                io::Error::new(
-                    io::ErrorKind::BrokenPipe,
-                    "the wasi pipe is not connected".to_string(),
-                )
-            })?;
+
+            // We need to figure out whether we need to block here.
+            // The problem is that in cases of multiple buffered reads like:
+            //
+            // println!("abc");
+            // println!("def");
+            //
+            // get_stdout() // would only return "abc\n" instead of "abc\ndef\n"
+
+            let data = match rx.try_recv() {
+                Ok(mut s) => {
+                    s.append(&mut rx.try_iter().flat_map(|f| f.into_iter()).collect());
+                    s
+                }
+                Err(_) => {
+                    // could not immediately receive bytes, so we need to block
+                    match rx.recv() {
+                        Ok(o) => o,
+                        // Errors can happen if the sender has been dropped already
+                        // In this case, just return 0 to indicate that we can't read any
+                        // bytes anymore
+                        Err(e) => {
+                            eprintln!("WasiPipe read error: {e}");
```

Comment:
It's not actually an error if the pipe disconnects (as the sender probably dropped the reference) so I don't think we should be logging this - besides logging to stderr  is not consistent with other logging locations which using the tracing crate.

Reply:
Resolved in https://github.com/wasmerio/wasmer/pull/3145/commits/238aef331b61b06a860dfacc13051fde7725c799